narwhal/types/src/consensus.rs

```diff
@@ -35,9 +36,39 @@ pub struct CommittedSubDag {
     pub sub_dag_index: SequenceNumber,
     /// The so far calculated reputation score for nodes
     pub reputation_score: ReputationScores,
+    /// The timestamp that should identify this commit. This is guaranteed to be monotonically
+    /// incremented. This is not necessarily the leader's timestamp. We compare the leader's timestamp
+    /// with the previously committed sud dag timestamp and we always keep the max.
+    pub commit_timestamp: TimestampMs,
 }
 
 impl CommittedSubDag {
+    pub fn new(
+        certificates: Vec<Certificate>,
+        leader: Certificate,
+        sub_dag_index: SequenceNumber,
+        reputation_score: ReputationScores,
+        previous_sub_dag: Option<&CommittedSubDag>,
+    ) -> Self {
+        let previous_sub_dag_ts = previous_sub_dag
+            .map(|s| s.commit_timestamp)
+            .unwrap_or_default();
+        let commit_timestamp = previous_sub_dag_ts.max(*leader.header().created_at());
+
+        if previous_sub_dag_ts > *leader.header().created_at() {
+            warn!(sub_dag_index = ?sub_dag_index, "Leader timestamp {} is older than previously committed sub dag timestamp {}. Auto-correcting to max {}.",
```

Comment:
This should probably be an `error`, because we do not expect this to happen with other conditions we put in proposer and vote process